pkg/controller/database/dbsubnetgroup/controller.go

```diff
@@ -18,175 +18,221 @@ package dbsubnetgroup
+		dbSubnetGroupclient, err := conn.newClientFn(ctx, []byte{}, p.Spec.Region, awsclients.UsePodServiceAccount)
+		return &external{client: dbSubnetGroupclient, kube: conn.kube}, errors.Wrap(err, errCreateDBSubnetGroupClient)
 	}
-	return &external{c}, nil
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := conn.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	dbSubnetGroupclient, err := conn.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: dbSubnetGroupclient, kube: conn.kube}, errors.Wrap(err, errCreateDBSubnetGroupClient)
 }
 
 type external struct {
-	client rds.DBSubnetGroupClient
+	client dbsg.Client
+	kube   client.Client
 }
 
-func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
-	cr, ok := mgd.(*v1alpha3.DBSubnetGroup)
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mgd.(*v1beta1.DBSubnetGroup)
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
 	}
 
 	req := e.client.DescribeDBSubnetGroupsRequest(&awsrds.DescribeDBSubnetGroupsInput{
-		DBSubnetGroupName: aws.String(cr.Spec.DBSubnetGroupName),
+		DBSubnetGroupName: aws.String(meta.GetExternalName(cr)),
 	})
-
-	response, err := req.Send(ctx)
-
+	res, err := req.Send(ctx)
 	if err != nil {
-		if rds.IsDBSubnetGroupNotFoundErr(err) {
-			return managed.ExternalObservation{
-				ResourceExists: false,
-			}, nil
-		}
-
-		return managed.ExternalObservation{}, errors.Wrapf(err, errDescribe, cr.Spec.DBSubnetGroupName)
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(dbsg.IsErrorNotFound, err), errDescribe)
 	}
 
 	// in a successful response, there should be one and only one object
-	if len(response.DBSubnetGroups) != 1 {
-		return managed.ExternalObservation{}, errors.Errorf(errMultipleItems, cr.Spec.DBSubnetGroupName)
+	if len(res.DBSubnetGroups) != 1 {
+		return managed.ExternalObservation{}, errors.Errorf(errZeroOrMoreResouce, meta.GetExternalName(cr))
 	}
 
-	observed := response.DBSubnetGroups[0]
-
-	cr.SetConditions(runtimev1alpha1.Available())
-
-	cr.UpdateExternalStatus(observed)
+	observed := res.DBSubnetGroups[0]
+	current := cr.Spec.ForProvider.DeepCopy()
+	dbsg.LateInitialize(&cr.Spec.ForProvider, &observed)
+	if !reflect.DeepEqual(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+	cr.Status.AtProvider = dbsg.GenerateObservation(observed)
+
+	switch cr.Status.AtProvider.State {
+	case v1beta1.DBSubnetGroupStateAvailable:
```

Comment:
I tested this locally and the status string I got for success was `Complete`, which does not appear in any of the state strings here. These strings look like copy of RDS state strings but every resource has its own set of state strings, so you either need to find them in AWS sdk or test locally to see what's returned.

Reply:
It's also fine to just have the success string (`Complete`) and let other states fall into `Unavailable` for now.

Reply:
I noticed that `DBSubnetGroup` does not have any specific status type as it's there for [RDS]. So while I was testing it I kept on receiving `Complete` state. Therefore, I have updated the code to just have the success string (`Complete`) and let other states fall into `Unavailable `as you suggested.

[RDS]: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.DBInstance.Status.html